Type hint for TypeConversionDict.get is too limited

Given the following code
```py
from werkzeug.datastructures import TypeConversionDict

d = TypeConversionDict()

def parse_int(value: str) -> int:
    return int(value)

d.get("foo", type=int)
d.get("foo", type=parse_int)
```

mypy complains with 
```log
foo.py:13:1: error: No overload variant of "get" of "TypeConversionDict" matches argument types "str", "Callable[[str], int]"  [call-overload]
    d.get("foo", type=parse_int)
    ^~~~~~~~~~~~~~~~~~~~~~~~~~~~
foo.py:13:1: note: Possible overload variants:
foo.py:13:1: note:     def get(self, key: Any) -> Any | None
foo.py:13:1: note:     def get(self, key: Any, default: Any) -> Any
foo.py:13:1: note:     def [T] get(self, key: Any, default: T) -> Any | T
foo.py:13:1: note:     def [T] get(self, key: str, type: type[T]) -> T | None
foo.py:13:1: note:     def [T] get(self, key: str, default: T, type: type[T]) -> T
```

It seems like that the type hint for `type` is too limited and should also include callables as the docstring says 

>         :param type: A callable that is used to cast the value in the
>                     :class:`MultiDict`.  If a :exc:`ValueError` or a
>                     :exc:`TypeError` is raised by this callable the default
>                     value is returned.

Environment:

- Python version: 3.12.0
- Werkzeug version: 3.1.1
- mypy version: 1.13.0

--- a/CHANGES.rst
+++ b/CHANGES.rst
@@ -5,6 +5,10 @@ Version 3.1.2
 
 Unreleased
 
+-   Improve type annotation for ``TypeConversionDict.get`` to allow the ``type``
+    parameter to be a callable. :issue:`2988`
+
+
 
 Version 3.1.1
 -------------

--- a/src/werkzeug/datastructures/headers.py
+++ b/src/werkzeug/datastructures/headers.py
@@ -114,11 +114,14 @@ class Headers(cabc.MutableMapping[str, str]):
     @t.overload
     def get(self, key: str, default: T) -> str | T: ...
     @t.overload
-    def get(self, key: str, type: type[T]) -> T | None: ...
+    def get(self, key: str, type: cabc.Callable[[str], T]) -> T | None: ...
     @t.overload
-    def get(self, key: str, default: T, type: type[T]) -> T: ...
+    def get(self, key: str, default: T, type: cabc.Callable[[str], T]) -> T: ...
     def get(  # type: ignore[misc]
-        self, key: str, default: str | T | None = None, type: type[T] | None = None
+        self,
+        key: str,
+        default: str | T | None = None,
+        type: cabc.Callable[[str], T] | None = None,
     ) -> str | T | None:
         """Return the default value if the requested data doesn't exist.
         If `type` is provided and is a callable it should convert the value,
@@ -153,15 +156,17 @@ class Headers(cabc.MutableMapping[str, str]):
             return rv
 
         try:
-            return type(rv)  # type: ignore[call-arg]
+            return type(rv)
         except ValueError:
             return default
 
     @t.overload
     def getlist(self, key: str) -> list[str]: ...
     @t.overload
-    def getlist(self, key: str, type: type[T]) -> list[T]: ...
-    def getlist(self, key: str, type: type[T] | None = None) -> list[str] | list[T]:
+    def getlist(self, key: str, type: cabc.Callable[[str], T]) -> list[T]: ...
+    def getlist(
+        self, key: str, type: cabc.Callable[[str], T] | None = None
+    ) -> list[str] | list[T]:
         """Return the list of items for a given key. If that key is not in the
         :class:`Headers`, the return value will be an empty list.  Just like
         :meth:`get`, :meth:`getlist` accepts a `type` parameter.  All items will
@@ -187,7 +192,7 @@ class Headers(cabc.MutableMapping[str, str]):
             for k, v in self:
                 if k.lower() == ikey:
                     try:
-                        result.append(type(v))  # type: ignore[call-arg]
+                        result.append(type(v))
                     except ValueError:
                         continue
 

--- a/src/werkzeug/datastructures/structures.py
+++ b/src/werkzeug/datastructures/structures.py
@@ -69,11 +69,14 @@ class TypeConversionDict(dict[K, V]):
     @t.overload
     def get(self, key: K, default: T) -> V | T: ...
     @t.overload
-    def get(self, key: str, type: type[T]) -> T | None: ...
+    def get(self, key: str, type: cabc.Callable[[V], T]) -> T | None: ...
     @t.overload
-    def get(self, key: str, default: T, type: type[T]) -> T: ...
+    def get(self, key: str, default: T, type: cabc.Callable[[V], T]) -> T: ...
     def get(  # type: ignore[misc]
-        self, key: K, default: V | T | None = None, type: type[T] | None = None
+        self,
+        key: K,
+        default: V | T | None = None,
+        type: cabc.Callable[[V], T] | None = None,
     ) -> V | T | None:
         """Return the default value if the requested data doesn't exist.
         If `type` is provided and is a callable it should convert the value,
@@ -108,7 +111,7 @@ class TypeConversionDict(dict[K, V]):
             return rv
 
         try:
-            return type(rv)  # type: ignore[call-arg]
+            return type(rv)
         except (ValueError, TypeError):
             return default
 
@@ -255,8 +258,10 @@ class MultiDict(TypeConversionDict[K, V]):
     @t.overload
     def getlist(self, key: K) -> list[V]: ...
     @t.overload
-    def getlist(self, key: K, type: type[T]) -> list[T]: ...
-    def getlist(self, key: K, type: type[T] | None = None) -> list[V] | list[T]:
+    def getlist(self, key: K, type: cabc.Callable[[V], T]) -> list[T]: ...
+    def getlist(
+        self, key: K, type: cabc.Callable[[V], T] | None = None
+    ) -> list[V] | list[T]:
         """Return the list of items for a given key. If that key is not in the
         `MultiDict`, the return value will be an empty list.  Just like `get`,
         `getlist` accepts a `type` parameter.  All items will be converted
@@ -279,7 +284,7 @@ class MultiDict(TypeConversionDict[K, V]):
         result = []
         for item in rv:
             try:
-                result.append(type(item))  # type: ignore[call-arg]
+                result.append(type(item))
             except (ValueError, TypeError):
                 pass
         return result
@@ -707,8 +712,10 @@ class _OrderedMultiDict(MultiDict[K, V]):
     @t.overload
     def getlist(self, key: K) -> list[V]: ...
     @t.overload
-    def getlist(self, key: K, type: type[T]) -> list[T]: ...
-    def getlist(self, key: K, type: type[T] | None = None) -> list[V] | list[T]:
+    def getlist(self, key: K, type: cabc.Callable[[V], T]) -> list[T]: ...
+    def getlist(
+        self, key: K, type: cabc.Callable[[V], T] | None = None
+    ) -> list[V] | list[T]:
         rv: list[_omd_bucket[K, V]]
 
         try:
@@ -720,7 +727,7 @@ class _OrderedMultiDict(MultiDict[K, V]):
         result = []
         for item in rv:
             try:
-                result.append(type(item.value))  # type: ignore[call-arg]
+                result.append(type(item.value))
             except (ValueError, TypeError):
                 pass
         return result
@@ -852,17 +859,20 @@ class CombinedMultiDict(ImmutableMultiDictMixin[K, V], MultiDict[K, V]):  # type
     @t.overload
     def get(self, key: K, default: T) -> V | T: ...
     @t.overload
-    def get(self, key: str, type: type[T]) -> T | None: ...
+    def get(self, key: str, type: cabc.Callable[[V], T]) -> T | None: ...
     @t.overload
-    def get(self, key: str, default: T, type: type[T]) -> T: ...
+    def get(self, key: str, default: T, type: cabc.Callable[[V], T]) -> T: ...
     def get(  # type: ignore[misc]
-        self, key: K, default: V | T | None = None, type: type[T] | None = None
+        self,
+        key: K,
+        default: V | T | None = None,
+        type: cabc.Callable[[V], T] | None = None,
     ) -> V | T | None:
         for d in self.dicts:
             if key in d:
                 if type is not None:
                     try:
-                        return type(d[key])  # type: ignore[call-arg]
+                        return type(d[key])
                     except (ValueError, TypeError):
                         continue
                 return d[key]
@@ -871,8 +881,10 @@ class CombinedMultiDict(ImmutableMultiDictMixin[K, V], MultiDict[K, V]):  # type
     @t.overload
     def getlist(self, key: K) -> list[V]: ...
     @t.overload
-    def getlist(self, key: K, type: type[T]) -> list[T]: ...
-    def getlist(self, key: K, type: type[T] | None = None) -> list[V] | list[T]:
+    def getlist(self, key: K, type: cabc.Callable[[V], T]) -> list[T]: ...
+    def getlist(
+        self, key: K, type: cabc.Callable[[V], T] | None = None
+    ) -> list[V] | list[T]:
         rv = []
         for d in self.dicts:
             rv.extend(d.getlist(key, type))  # type: ignore[arg-type]

